tgui/packages/tgui/interfaces/Fabricator.tsx

```diff
@@ -0,0 +1,337 @@
+import { useBackend, useSharedState } from '../backend';
+import { Stack, Section, Button, Input, Icon, Tabs, Dimmer } from '../components';
+import { Window } from '../layouts';
+import { Material, MaterialAmount, MaterialFormatting, Materials, MATERIAL_KEYS } from './common/Materials';
+import { Fragment } from 'inferno';
+
+type Design = {
+  name: string;
+  desc: string;
+  cost: { [K in keyof typeof MATERIAL_KEYS]: number };
+  id: string;
+  categories?: string[];
+  maxstack: number;
+  icon: string;
+};
+
+type FabricatorData = {
+  materials: Material[];
+  fab_name: string;
+  on_hold: boolean;
+  designs: { [K: string]: Design };
+  busy: boolean;
+};
+
+type AvailableMaterials = { [K in keyof typeof MATERIAL_KEYS]?: number };
+
+const SearchBar = (props, context) => {
+  const [searchText, setSearchText] = useSharedState(
+    context,
+    'search_text',
+    ''
+  );
+
+  return (
+    <Stack align="baseline">
+      <Stack.Item>
+        <Icon name="search" />
+      </Stack.Item>
+      <Stack.Item grow>
+        <Input
+          fluid
+          placeholder="Search for..."
+          onInput={(_e: unknown, v: string) => setSearchText(v.toLowerCase())}
+          value={searchText}
+        />
+      </Stack.Item>
+    </Stack>
+  );
+};
+
+export const Fabricator = (props, context) => {
+  const { act, data } = useBackend<FabricatorData>(context);
+  const { materials, fab_name, on_hold, designs, busy } = data;
+  const sortedDesigns = Object.values(designs).sort((a, b) =>
+    a.name > b.name ? 1 : 0
+  );
+  const categoryCounts: { [K: string]: number } = {};
+  const [selectedCategory, setSelectedCategory] = useSharedState(
+    context,
+    'selected_category',
+    '__ALL'
+  );
+  const [searchText, setSearchText] = useSharedState(
+    context,
+    'search_text',
+    ''
+  );
+  const [displayMatCost, setDisplayMatCost] = useSharedState(
+    context,
+    'display_material_cost',
+    true
+  );
+
+  let recipeCount = 0;
+
+  Object.values(sortedDesigns).map((design) => {
```

Comment:
Just make these standard for loops instead. `map` should not be performing side effects.

Reply:
I have no idea where `forEach` went, but I miss it...